### ISSUE TITLE
Cast with user-defined conversion doesn't honor [AllowNull].

### ISSUE THREAD

**Issue by Yen-David** [dotnet/roslyn#40113]
**Version Used**: C# 8, .Net Standard 2.1, Nullable Reference Context enabled

**Steps to Reproduce**:

1. Create a generic wrapper for all variables, fields, parameters and returns, to protect them from future syntax changes in C#, such as J{Y} where Y is anything and everything.
2. Add implicit casts to make everything transparent, in and out of code, new and old, however interpolated and strung together.
3. I still get compiler warnings!

**Expected Behavior**:

I must not get compiler warnings.

**Actual Behavior**:

I get compiler warnings.

```
 [Serializable]
 public readonly partial struct J<Y>
 {
  public static implicit operator J<Y>([AllowNull]Y x) => new J<Y>(x);
  internal J([AllowNull]Y y) => _y = y;
  [MaybeNull] private Y _y { get; }
 }
 partial struct JHopper<Y> where Y : class
 {
  private Y? _q;
  public J<Y> ZHop() => _q; // <= NULLABILITY WARNING HERE?! [AllowNull] isn't honored, obviously.
 }
```
[sharplab](https://sharplab.io/#v2:D4AQDABCCMB0AiBLAhgcwHYHsDOAXRAxtrAMKYAmApgILrIA2AntotgNwCwAUNwMToBXevWQAjepQiU64yt25QAzBABOlZOUzomEAA7IV+BhDwqBBXBABSAHgCaAPgUBvBUqjQAbBEQBbXfSEiJaYupQqyLiYKtb2DgAUANrUwpgA7gByQvQAunYQAB4AlBAAvA4Q6JRpsY7xxZxcED7ouOF09NZJKfTpWcJ5EIwl5RAA+oxlQ43NiQCyyIyilP25eiqIAG6RkvkTEM4QqJS4bBAAvgqXTfqGKJ2m5pZWABKhYSpxEGkAFuG7EAAXFAAEwuNy6DbbNoQOwAfnGAEcZu5bI4IAAtN66eIjCpjZEQAD0RIgNlKEAyAFUADI06gAIQAkjSmQAVfIAdWoACUMkyMgBxCAvACiPNFcIAhBBkqlMtkcj5sOgAOSWH5aaKUcgAGggmFEm0QmAE2CYsCuQA=)

Why is this valuable? Well, C# can change syntax and semantics however it wants, but all my variables, fields, parameters and returns will ALWAYS be protected in J{Y}, saving them from most possible future rewrites. Performance is fairly good, as short methods are ALWAYS inlined, the best of both worlds!

**Comment by RikkiGibson** [comment]
I don't see the declaration of the field `_q` in the above sample. Could you please try to reproduce the problem with a more minimal sample?

**Comment by Yen-David** [comment]
@RikkiGibson, just added, thanks!

**Comment by sharwell** [comment]
I tend to use the backwards compatibility policy of C# as protection against future breaking changes. I find it a bit easier than manually wrapping everything, and it doesn't require that all my wrappers get wrapped to avoid breaking changes in the wrappers themselves. 😄 

This issue is a specific case from #36242 (`AllowNull` on input to conversion operator).

**Comment by Yen-David** [comment]
@sharwell, I have read the related issues, from way back in May & June. May you add this one to the list as well? Operator overloading is an important feature in C#, which many and I use very often. Wrappers are great due to Find & Replace in Files with Visual Studio's regex support. The nullability modifier "?" cannot be inferred automatically because it's a semantic thing. If I just wrap all variables, fields, parameters and returns, it's a syntactic thing!

**Comment by jcouv** [comment]
I'll go ahead and close this issue as duplicate of https://github.com/dotnet/roslyn/issues/32671 (Annotations do not work for user defined conversion operators)

**Comment by Yen-David** [comment]
@jcouv, no problem. Hopefully, we'll get the fix soon, maybe in Visual Studio 16.5!
